### PR TITLE
Kotlin dsl support

### PR DESCRIPTION
**Pull request by marcoferrer**
~~This is not ready to merge and is still a work in progress but I wanted to open the pr a little early to get some feedback on the proposed approach to resolving  #219.~~

### CODE REVIEW

**Comment by zpencer** [comment]
dummy comment to try and unwedge CLA bot

**Comment by zpencer** [comment]
I like this idea of adding extensions functions, and then changing the groovy API only if extensions can not solve the problem. Taking a compile only dependency on kotlin is fine.

Would it be possible to add a test case that shows how a kotlin DSL build file looks? https://github.com/google/protobuf-gradle-plugin/pull/205/files may be a possible starting point.

File: exampleProject/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    compile("io.grpc:grpc-protobuf:1.0.0-pre2")
+    if (JavaVersion.current().isJava9Compatible) {
+        // Workaround for @javax.annotation.Generated
+        // see: https://github.com/grpc/grpc-java/issues/3633
+        compile("javax.annotation:javax.annotation-api:1.3.1")
+    }
+    // Extra proto source files besides the ones residing under
+    // "src/main".
+    "protobuf"(files("lib/protos.tar.gz"))
+    "protobuf"(fileTree("ext/"))
```
Comment (marcoferrer) [inline exampleProject/build.gradle.kts]: using string configuration access but i should probably look into adding an extension scoped to the dependency handler

File: build.gradle [hunk 2]
```diff
@@ -222,3 +228,11 @@ if (System.env.TRAVIS == 'true') {
     }
   }
 }
+
+// Temporary work around to get compilation working during development
+// Considering moving to a separate module
+// https://discuss.gradle.org/t/kotlin-groovy-and-java-compilation/14903/10
+compileGroovy.dependsOn = compileGroovy.taskDependencies.mutableValues - 'compileJava'
```
Comment (marcoferrer) [inline build.gradle]: This is really ugly so Im hoping to move it to a seperate module. Im just not sure where it should live in the project.

**Comment by marcoferrer** [comment]
@zpencer I added an example of a kotlin dsl script based on the original example. It is still in the example directory while I try to figure out how to wire up a test project for this that would work with the build process. Im open to suggestions on how to approach it.

**Comment by marcoferrer** [comment]
Tested the current impl in the new idea project and everything seems to be working so far.
Type resolution and code generation are working as expected.

screen shot 2018-08-23 at 11 38 15 pm

**Comment by gmale** [comment]
This is a huge improvement over the currently required syntax!

I'm not sure if this question is out of scope (i.e. too complex of a change to the underlying plugin). I only bring it up since the PR requested feedback: 

Is there a way to reduce the amount of boilerplate? All the developer is really trying to convey [here](https://github.com/google/protobuf-gradle-plugin/pull/262#issuecomment-415641784) is, "I want to use protobuf with grpc and javalite" which is probably an extremely common usecase for `kotlin-dsl` users. Any logic that could be moved out of the extension object and into the plugin would make life easier for many users. For example:

```
protobuf {
    plugins {
        id("grpc") option "lite"
    }
}
```

**Comment by marcoferrer** [comment]
Performed some clean up and pushed the updated example project. 

I think there might be a way to implement the shorter syntax for options in the task.plugin config block. ie
```
ofSourceSet("main").forEach{
            it.plugins {
                id("grpc") option "testing"
            }
}
```


Im just not sure having options configurable in the protobuf.plugins block is a good idea since the assumed config would be implicitly applied to all  source sets.

**Comment by marcoferrer** [comment]
@zpencer Im not sure if my build is failing to due master failing?

File: src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+): ExternalModuleDependency =
+    add("protobuf", create(group, name, version, configuration, classifier, ext), dependencyConfiguration)
+
+inline fun <T : ModuleDependency> DependencyHandler.protobuf(dependency: T, dependencyConfiguration: T.() -> Unit): T =
+    add("protobuf", dependency, dependencyConfiguration)
+
+/**
+ * The 'testProtobuf' configuration.
+ */
+val ConfigurationContainer.testProtobuf: Configuration
```
Comment (zpencer) [inline src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt]: Can this be more general? It looks like this gives special treatment to the main sourceset (`protobuf`) and the test sourceset (`testProtobuf`) but won't handle user created sourcesets like this:


```groovy
sourceSets {
    main {
    }
    foo {
        java {
            srcDirs 'src/foo/'
        }
    }
}


dependencies {
  fooProtobuf files("/tmp/protos-test.tar.gz")
}
```

`createConfiguration` is where the plugin creates the `xxxxProtobuf` configurations in the project: 
https://github.com/google/protobuf-gradle-plugin/blob/master/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy#L154

Maybe there is some way this extension can plug into them in a general way.

File: src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+import org.gradle.api.artifacts.dsl.DependencyHandler
+import org.gradle.kotlin.dsl.*
+
+
+fun Project.protobuf(configuration: ProtobufConfigurator.() -> Unit) {
+    project.convention.getPlugin(ProtobufConvention::class.java).protobuf.apply(configuration)
+}
+
+fun ProtobufConfigurator.protoc(closure: ExecutableLocator.() -> Unit) {
+    protoc(delegateClosureOf(closure))
```
Comment (zpencer) [inline src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt]: I am still trying to wrap my head around `delegateClosureOf` vs `closureOf` (https://github.com/gradle/kotlin-dsl/blob/master/doc/getting-started/Closures.md). Is `delegateClosureOf` meant for closures that make use of the [closure delegate from groovy](http://groovy-lang.org/closures.html#closure-owner) ? As far as I can tell this plugin does not use this feature.

File: src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    })
+}
+
+fun GenerateProtoTask.plugins(closure: NamedDomainObjectContainerScope<GenerateProtoTask.PluginOptions>.() -> Unit) {
+    plugins(delegateClosureOf<NamedDomainObjectContainer<GenerateProtoTask.PluginOptions>> {
+        this{ this.closure() }
+    })
+}
+
+fun ProtobufConfigurator.GenerateProtoTaskCollection.ofSourceSet(sourceSet: String): Collection<GenerateProtoTask> =
```
Comment (zpencer) [inline src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt]: Why can't the groovy version of these functions be called directly? Maybe a comment would be helpful.

**Comment by marcoferrer** [comment]
> I'm still learning Kotlin, so please excuse my basic questions about syntax that is hard to learn about by Googling :)

Yeah I'm in a similar boat with groovy. Feel free to ask any questions even if they arent related to the pr. The only curve ball in regards to the kotlin code in the pr is that most kotlin gradle dsl scripts dont follow normal kotlin conventions. The dsl has a lot of extra functions and helpers to aid in shimming dynamic groovy into static kotlin. Ill address your questions and concerns now.

File: src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    project.convention.getPlugin(ProtobufConvention::class.java).protobuf.apply(configuration)
+}
+
+fun ProtobufConfigurator.protoc(closure: ExecutableLocator.() -> Unit) {
+    protoc(delegateClosureOf(closure))
+}
+
+fun ProtobufConfigurator.plugins(closure: NamedDomainObjectContainerScope<ExecutableLocator>.() -> Unit) {
+    plugins(delegateClosureOf<NamedDomainObjectContainer<ExecutableLocator>> {
+        this{ this.closure() }
```
[the remark this answers is not in the document]
Comment (marcoferrer) [reply src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt]: Yeah looks gross and after doing some additional digging I updated with a cleaner usage. This implementation was a port from the existing example provided here https://github.com/google/protobuf-gradle-plugin/issues/219#issuecomment-376179475

Back to your original question. There are a couple of Kotlin features at work here that make up this statement.
 * [Lambda with Receiver](https://kotlinlang.org/docs/reference/lambdas.html#function-literals-with-receiver)
   * The ```closureOf``` method takes a lambda and the ```this``` property in that lambda is the argument passed in.
 * [Invoke Operator Overloading](https://kotlinlang.org/docs/reference/operator-overloading.html#invoke)
   * The gradle kotlin dsl provides and extension function overloading the [invoke](https://github.com/gradle/kotlin-dsl/blob/master/subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt#L34) operator for objects with type ```NamedDomainObjectContainer```
   * This function allows us to apply our configuration to the receiver
 * [Lambda as Last Param](https://kotlinlang.org/docs/reference/lambdas.html#passing-a-lambda-to-the-last-parameter)
   * If the last argument to a function is a lambda then the parenthesis can be omitted.

In the latest commit, this code has been refactored to something thats a little more readable, after looking into what exactly the method needed to achieve.
``` 
fun GenerateProtoTask.plugins(closure: NamedDomainObjectContainerScope<GenerateProtoTask.PluginOptions>.()-> Unit) {
    plugins(closureOf<NamedDomainObjectContainer<GenerateProtoTask.PluginOptions>> {
        this(closure)
    })
}
```

File: src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt (hunk 3 above)
Comment (marcoferrer) [reply src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt]: So I cant dynamically provide the dependency configuration methods, but I can try providing a nicer mechanism. 

You can try it out in the latest commit.
```
dependencies{
    protobuf["sample"]("com.google.protobuf:protobuf-java:3.6.1")
}
```

File: src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt (hunk 4 above)
Comment (marcoferrer) [reply src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt]: So is a by-product of my ignorance of groovy. I updated to use ```closureOf```

**Comment by marcoferrer** [comment]
I added support for configuring sourceSets. One of the last things is sorting out setting up a test project so that the kotlin extensions are backed by tests and not just the example project.

File: src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt (hunk 5 above)
Comment (marcoferrer) [reply src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt]: I added a comment to the source but Ill post it here as well.

The method ```generatorProtoTasks``` applies the supplied closure to an
instance of ```GenerateProtoTaskCollection```.

Since ```JavaGenerateProtoTaskCollection``` and ```AndroidGenerateProtoTaskCollection```
each have unique methods, and only one instance in allocated per project, we need a way to statically resolve the available methods. This is a necessity since Kotlin does not have any dynamic method resolution capabilities.

File: src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt (hunk 3 above)
Comment (marcoferrer) [reply src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt]: This solution has some pros and cons so I wanted to discuss them with you when possible. Ill add a full explanation in a few.

File: src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt (hunk 5 above)
Comment (zpencer) [reply src/main/kotlin/com/google/protobuf/gradle/KotlinInteroperability.kt]: Ah, so the problem is the dynamic typing of `GenerateProtoTaskCollection`, since we don't know whether it is `JavaGenerateProtoTaskCollection` or `AndroidGenerateProtoTaskCollection`, kotlin's static typing means we can call neither's methods. 

What do you think about adding some extension methods to `GenerateProtoTaskCollection` like `asJavaTasks` returning the same object casted as `JavaGenerateProtoTaskCollection` and `asAndroidTasks` casting as `AndroidGenerateProtoTaskCollection`. That way, we minimize the amount of extension code that must be kept in sync with the groovy DSL.

**Comment by marcoferrer** [comment]
Regarding the dependency handling extensions I have some points I want to highlight.

So originally I needed to add extensions for the ```protobuf``` and ```testProtobuf``` to the build script dependencies block. It turns out that the extensions are only needed when you're applying a plugin using the legacy syntax. 

If you use the gradle ```plugins{ }``` block then gradle will generate synthetic extensions for you in a file called ```ConfigurationAccessors.kt```. 

[Gradle Docs: Configurations and Dependencies](https://guides.gradle.org/migrating-build-logic-from-groovy-to-kotlin/#configurations-and-dependencies)

It looks like they only generate the methods for the default source sets. Dependencies for custom source sets would still needed to be added via the common convention:
```
val sampleProtobuf by configurations

dependencies{
     sampleProtobuf(......)
}
```
 
So this leaves us at a cross roads. Do we provide conveniences for users so that it doesnt matter what style they use to apply the plugin, or do we follow the default gradle conventions? For the default source sets I can see it being ok to explicitly provide the methods. Its only weird when we look at custom source sets and introduce a custom convention specific to this plugin.
```
dependencies{
    protobuf["sample"]("com.google.protobuf:protobuf-java:3.6.1")
}
```

What do you think?

**Comment by zpencer** [comment]
Looking at `accessors.kt` that comes with the kotlin DSL, I see a precedent for special casing a few common configurations like `compile` and `testCompile` for convenience. So now think it's fine to special case `protobuf` and `testProtobuf`.

As for the custom source sets, maybe we don't need to do anything at all. After experimenting with the kotlin DSL, I realized that this standard syntax should just work. 
```kotlin
dependencies {
    add("fooProtobuf", files("/tmp/protos-test.tar.gz"))
}
```
This doesn't look too awful and we can probably simplify the current ProtobufDependencyConfiguration.kt with this approach.

**Comment by marcoferrer** [comment]
Awesome Ill clean up ProtobufDependencyConfiguration.kt. 

In regards to writing up tests, Im still trying to get a test project working with the existing test classes. There isnt much available in terms of resources for testing kotlin dsl projects using the gradle runner so its been a slow process. Just wanted to give a heads up.

**Comment by marcoferrer** [comment]
Also there are extensions built into the kotlin gradle dsl that clean up the custom source set syntax a little further
```
dependencies {
    "fooProtobuf"(files("/tmp/protos-test.tar.gz"))
}
```

File: src/main/kotlin/com/google/protobuf/gradle/ProtobufDependencyConfiguration.kt [hunk 7]
```diff
@@ -0,0 +1,86 @@
+package com.google.protobuf.gradle
+
+import org.gradle.api.artifacts.*
```
[the remark this answers is not in the document]
Comment (marcoferrer) [reply src/main/kotlin/com/google/protobuf/gradle/ProtobufDependencyConfiguration.kt]: Fixed

File: testProjectKotlinDslBase/build.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+            }
+            file = file.parentFile
+        }
+    }
+    assert(expectedDirs == actualDirs)
+}
+
+// This extension is not auto generated when we apply the plugin using
+// apply(plugin = "com.google.protobuf")
+val Project.protobuf: ProtobufConvention get() =
```
Comment (marcoferrer) [inline testProjectKotlinDslBase/build.gradle.kts]: Since this is another extension that relies on the plugin being applied in the plugins block, should we add it to ```ProtobufConfiguratorExts.kt``` as well?

**Comment by marcoferrer** [comment]
Finally got tests added. PR is ready for a full review whenever you get a chance.

File: examples/exampleKotlinDslProject/settings.gradle.kts [hunk 9]
```diff
@@ -0,0 +1,15 @@
+pluginManagement{
+    repositories {
+        mavenLocal()
```
[the remark this answers is not in the document]
Comment (marcoferrer) [reply examples/exampleKotlinDslProject/settings.gradle.kts]: Try running ```./gradlew install -x test```

File: testProjectKotlinDslBase/build.gradle.kts (hunk 8 above)
Comment (marcoferrer) [reply testProjectKotlinDslBase/build.gradle.kts]: This extension is used to get the plugin convention from the project, which was needed for Line 106. I didnt realize that ```HasConvention``` isnt part of the public api. The code itself I copied from the dynamically generated code file that gradle creates when you use the plugins block. I can see if there's a way around casting to ```HasConvention```

File: examples/exampleKotlinDslProject/settings.gradle.kts (hunk 9 above)
[the remark this answers is not in the document]
Comment (marcoferrer) [reply examples/exampleKotlinDslProject/settings.gradle.kts]: Also the example project needs at least gradle 4.10 to run. For two reasons, 1) Its the version that includes the kotlin dsl rc, 2) First version to allow snapshot ids in the plugins / pluginsManagement blocks.

File: testProjectKotlinDslBase/build.gradle.kts (hunk 8 above)
Comment (marcoferrer) [reply testProjectKotlinDslBase/build.gradle.kts]: Cleaned up the usage to the following
```
val Project.protobuf: ProtobufConvention get() =
    this.convention.getPlugin(ProtobufConvention::class)
```

File: gradle/wrapper/gradle-wrapper.properties [hunk 10]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.8-bin.zip
```
Comment (zpencer) [inline gradle/wrapper/gradle-wrapper.properties]: undo this change

File: examples/exampleKotlinDslProject/settings.gradle.kts (hunk 9 above)
[the remark this answers is not in the document]
Comment (zpencer) [reply examples/exampleKotlinDslProject/settings.gradle.kts]: Let's keep the md5 of the gradlew files in sync with the top level project.

File: testProjectKotlinDslBase/build.gradle.kts (hunk 8 above)
Comment (zpencer) [reply testProjectKotlinDslBase/build.gradle.kts]: nit: Can this workaround be moved nearer to `apply(plugin = "com.google.protobuf")` ?

File: examples/exampleKotlinDslProject/settings.gradle.kts (hunk 9 above)
[the remark this answers is not in the document]
Comment (marcoferrer) [reply examples/exampleKotlinDslProject/settings.gradle.kts]: I feel pretty dumb but Im guessing you mean make sure that these gradlew files are exactly the same as the ones in the root project?

[the remark this answers is not in the document]
Comment (marcoferrer) [reply examples/exampleKotlinDslProject/settings.gradle.kts]: Everything looks good now

[the remark this answers is not in the document]
Comment (zpencer) [reply examples/exampleKotlinDslProject/settings.gradle.kts]: Yes that was what I meant, I should have been more clear :)

[the remark this answers is not in the document]
Comment (marcoferrer) [reply examples/exampleKotlinDslProject/settings.gradle.kts]: I think its my local git settings messing it up. Let me try again

[the remark this answers is not in the document]
Comment (marcoferrer) [reply examples/exampleKotlinDslProject/settings.gradle.kts]: Fixed

**Comment by zpencer** [comment]
@zhangkun83 would you like to take a look before I merge?

**Comment by zhangkun83** [comment]
I took a glance and the DSL looks pretty good. I trust @zpencer's judgement. Good job guys.

**Comment by zpencer** [comment]
Great! Let's merge it so our users can start using this experimental feature.

**Comment by marcoferrer** [comment]
Really happy to see this get merged. @zpencer Thanks again for all your help. When I get some free time Im hoping to add more documentation to the new code and possibly some additional tests.

**Comment by zhangkun83** [comment]
0.8.7 is released!
